javascript/diff_match_patch_uncompressed.js

```diff
@@ -2215,6 +2215,46 @@ diff_match_patch.prototype.patch_splitMax = function(patches) {
   }
 };
 
+diff_match_patch.prototype.diffs_joinSurrogatePairs = function(diffs) {
+	var lastEnd;
+	var overwrittenDiffsCounter = 0;
+	
+	for (var x = 0 ; x < diffs.length ; x++) {
+		var thisDiff = diffs[x];
+		var thisTop = thisDiff[1][0];
+		var thisEnd = thisDiff[1][thisDiff[1].length - 1];
+		
+		if (0 === thisDiff[1].length) {
+			continue;
+		}
+		
+		// trap a trailing high-surrogate so we can
+		// distribute it to the successive edits
+		if (thisEnd && this.isHighSurrogate(thisEnd)) {
+			lastEnd = thisEnd;
+			thisDiff[1] = thisDiff[1].slice(0, -1);
+		}
+		
+		if (lastEnd && thisTop && this.isHighSurrogate(lastEnd) && this.isLowSurrogate(thisTop)) {
+			thisDiff[1] = lastEnd + thisDiff[1];
+		}
+		
+		if (0 === thisDiff[1].length) {
+			continue;
+		}
+		
+		diffs[overwrittenDiffsCounter] = thisDiff;
+		overwrittenDiffsCounter ++;
+	}
+	
+	return diffs.splice(0, overwrittenDiffsCounter)
```

Comment:
can you help me understand what's going on here with `overwrittenDiffsCounter`? at a glance it looks like we're overwriting the diff operations starting at the first one and then removing those all at the end.

in my head this should be scanning through the array and potentially modifying the current and previous diff objects, but not removing the front of the list of them.

Reply:
if it's not easy to do this in-place then I suppose bringing back the array cloning, as you initially proposed, might be a practical solution 🤷‍♂️